core/src/main/java/org/wso2/msf4j/internal/MSF4JMessageProcessor.java

```diff
@@ -167,7 +256,275 @@ public void setTransportSender(TransportSender transportSender) {
+        Object endpoint = routableEndpoint.getDestination();
+        Map<String, String> paramValues = routableEndpoint.getGroupNameValues();
+        Method method = new EndpointDispatcher().getOnStringMessageMethod(endpoint);
+        try {
+            List<Object> parameterList = new LinkedList<>();
+            boolean isStringSatifsfied = false;
+            Arrays.stream(method.getParameters()).forEach(
+                    parameter -> {
+                        if (parameter.getType() == String.class) {
+                            PathParam pathParam = parameter.getAnnotation(PathParam.class);
+                            if (pathParam == null) {
+                                parameterList.add(textCarbonMessage.getText());
+                            } else {
+                                if (isStringSatifsfied == false) {
+                                    parameterList.add(paramValues.get(pathParam.value()));
+                                } else {
+                                    throw new IllegalArgumentException("String parameters without" +
+                                                                                     " @PathParam annotation");
+                                }
+                            }
+                        } else if (parameter.getType() == Session.class) {
+                            parameterList.add(session);
+                        } else {
+                            parameterList.add(null);
+                        }
+                    }
+            );
+            executeMethod(method, endpoint, parameterList, session);
+        } catch (Throwable throwable) {
+            handleError(textCarbonMessage, throwable, routableEndpoint, session);
+        }
+    }
+
+    /*
+    Handle Binary WebSocket Message
+     */
+
+    private void handleBinaryWebSocketMessage(BinaryCarbonMessage binaryCarbonMessage,
+                                              PatternPathRouter.RoutableDestination<Object>
+                                                      routableEndpoint, Session session) {
+        Object webSocketEndpoint = routableEndpoint.getDestination();
+        Map<String, String> paramValues = routableEndpoint.getGroupNameValues();
+        Method method = new EndpointDispatcher().getOnBinaryMessageMethod(webSocketEndpoint);
+        try {
+            List<Object> parameterList = new LinkedList<>();
+            Arrays.stream(method.getParameters()).forEach(
+                    parameter -> {
+                        if (parameter.getType() == ByteBuffer.class) {
+                            parameterList.add(binaryCarbonMessage.readBytes());
+                        } else if (parameter.getType() == byte[].class) {
+                            ByteBuffer buffer = binaryCarbonMessage.readBytes();
+                            byte[] bytes = new byte[buffer.capacity()];
+                            for (int i = 0; i < buffer.capacity(); i++) {
```

Comment:
Why are you using capacity() here? I believe you only need to copy the valid bytes. and there is a bulk get[1] method for that.
ByteBuffer buf = ...
byte[] arr = new byte[buf.remaining()];
buf.get(arr);

[1] https://docs.oracle.com/javase/7/docs/api/java/nio/ByteBuffer.html#get(byte[])